### DISCUSSION TITLE
llama-cli CPU Control | Pin to Physical Cores

### DISCUSSION THREAD

**Discussion by Allan-Luu** [ggml-org/llama.cpp#9996] [Q&A] [answered]
Hello, I'm trying to run llama-cli and pin the load onto the physical cores of my CPUs. On my processors, I have 128 physical cores and I want to run some tests on maybe the first 0-8, then 0-16, then 0-24, and so on adding 8 each time all the way until I get to using all 128 physical cores.

I'm trying to run this line:

`./llama-cli -m /data/models/llama-2-7b.Q4_K_M.gguf -p "Tell me a joke" -n 8 --numa distribute -Crb 0-7 -Cb 0xFF --cpu-strict-batch 1`

But when running `htop` I see most of the cores are lighting up, which is not my intention. I only want to see the processes run on the first 8 cores in this case. 

![image](https://github.com/user-attachments/assets/c36ea927-317d-49d5-bbb2-77d39a87521a)
Screenshot of htop CPU load when running the line of code above.  

Would someone be able to help me figure out how to pin the load to the exact cores of the CPU?

**Comment by ggerganov** [comment]
Pinging @bmtwl for any insights

**Comment by bmtwl** [comment, marked as answer]
You're going to want to use numactl to give you fine-grained control over where the threads execute:
`numactl -N0-7 -m0-7 /path/to/llama-cli -m /path/to/model.gguf -t 8 --numa numactl`
You can check the layout of your system with `numactl -H` so that you're getting the behaviour you want when assigning numa nodes

**Reply by Allan-Luu** [reply]
Hey @bmtwl , do you know whether we have control of the `split mode` across CPUs? Similar to the `--split-mode` for GPUs?

**Comment by max-krasnyansky** [comment]
When we added the threadpool and the new `--cpu-mask/range/strinct` options we tried to avoid messing with the numa distribute logic.
So currently those two options (ie using both `--numa distribute` and `--cpu-mask / --cpu-strict`) are not compatible. 
I'm going to follow up on this in the next round of threading updates (been meaning to work on that but keep getting distracted with other stuff).

@Allan-Luu based on your description above you don't even need to specify the `--numa` option.
Just use the `--cpu-mask\range\strict` options.

`./llama-cli -m /data/models/llama-2-7b.Q4_K_M.gguf -p "Tell me a joke" -n 8 --cpu-range 0-7 --cpu-strict 1`
 
 This will strictly bind the threads to CPU cores 0-7, one thread per core.

**Reply by Allan-Luu** [reply]
thank you all for your help!

hmm, @max-krasnyansky when running the command in your response, it uses a thread count of 128 instead of 8 and uses all cores on the CPUs

```
./llama-cli -m /data/models/llama-2-7b.Q4_K_M.gguf -p "Tell me a joke" -n 8 --cpu-range 0-7 --cpu-strict 1
```
![image](https://github.com/user-attachments/assets/6161e00a-0c95-42ea-8064-15b2f6bb9885)
![image](https://github.com/user-attachments/assets/95f96b3b-2941-4766-a177-2654503c6e32)
![image](https://github.com/user-attachments/assets/90b3ceab-d4d1-43fb-a985-8da026c222e7)



but when running with the suggestion from @bmtwl it works! I had to use `-N0-1` to specify the 2 processors on my machine (2 chips), `-C0,1,2,3,4,5,6,7` is specifying each CPU I intend to use, and `-t 8` determines the number of threads I want for this run.
```
numactl -N0-1 -C0,1,2,3,4,5,6,7 ./llama-cli -m /data/models/llama2/llama-2-13b.Q2_K.gguf -t 8 --numa numactl
```

![image](https://github.com/user-attachments/assets/16d2d73b-2485-4666-90ee-17c937a1820e)
![image](https://github.com/user-attachments/assets/7c3d599a-542d-458f-a9a1-2b9e8aa6fae3)

Is there any reason not to use `numactl`?

**Reply by max-krasnyansky** [reply]
Oh. Sorry. My bad. A little typo `-n` instead of `-t`. 

`./llama-cli -m /data/models/llama-2-7b.Q4_K_M.gguf -p "Tell me a joke" -t 8 --cpu-range 0-7 --cpu-strict 1`

It's fine to use `numactl` in general but `--cpu-mask\range\strict` provide more precise control 
(strict one thread per core, different masks for batch/non-batch, etc).

**Reply by MSZHabibie** [reply]
May I know why you included --numa numactl at the end? I tried running it without that, and it worked fine. BTW thanks for this post i can finally limit llama.cpp load model cpu core usage

**Reply by Allan-Luu** [reply]
I was using `--numa numactl` at the end based on bmtwl's answer above, didn't really consider whether to exclude it or not, but it's great to know that it works without it! And I'm glad my question was helpful to you.

And thanks for the explanation @max-krasnyansky I'll do some more testing when I get a chance to compare the two to see which one is a better fit for me.
